can-io-firmware/src/servo.c

```diff
@@ -0,0 +1,93 @@
+#include <string.h>
+
+#include "pwm.h"
+#include "servo.h"
+#include "quadramp/quadramp.h"
+
+#define SERVO_PWM_TIMER_FREQ    1000000 // 1MHz
+#define SERVO_PWM_PERIOD        20000   // 20 ms period
+
+typedef struct {
+    float setpoint; // Duty cycle setpoint
+    struct quadramp_filter filter;
+} servo_t;
+
+static servo_t servo[4];
+
+/* convert 0.0-1.0 to full pwm duty cycle. */
+static uint32_t duty_cycle(float pos)
+{
+    if (pos > 1) {
+        pos = 1;
+    } else if (pos < 0) {
+        pos = 0;
+    }
+    return (uint32_t)(pos * SERVO_PWM_TIMER_FREQ);
+}
+
+/* Not set if coefficient is too small (close to zero) */
+static void set_quadramp_coefficients(struct quadramp_filter* filter, float vel, float acc)
+{
+    if (vel > 1e-7) {
+        quadramp_set_1st_order_vars(filter, vel * SERVO_PWM_TIMER_FREQ, vel * SERVO_PWM_TIMER_FREQ);
+    }
+    if (acc > 1e-7) {
+        quadramp_set_2nd_order_vars(filter, acc * SERVO_PWM_TIMER_FREQ, acc * SERVO_PWM_TIMER_FREQ);
+    }
+}
+
+void servo_set(const float pos[4], const float vel[4], const float acc[4])
+{
+    servo[0].setpoint = pos[0];
```

Comment:
Please add proper locking since `servo` is shared between the UAVCAN thread and the servo_thd.

Reply:
Good catch!